File: src/main/java/com/github/jactor/web/consumer/UserConsumer.java
package com.github.jactor.web.consumer;

import com.github.jactor.shared.dto.UserDto;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

public class UserConsumer {

  private final RestTemplate restTemplate;

  public UserConsumer(RestTemplate restTemplate) {
    this.restTemplate = restTemplate;
  }

  public Optional<UserDto> find(String username) {
    ResponseEntity<UserDto> responseEntity = restTemplate.getForEntity("/user/name/" + username, UserDto.class);

    return Optional.ofNullable(bodyOf(responseEntity));
  }

  public List<String> findAllUsernames() {
    ResponseEntity<String[]> responseEntity = restTemplate.getForEntity(
        "/user/active/usernames", String[].class
    );

    return List.of(Objects.requireNonNull(responseEntity.getBody()));
  }

  private UserDto bodyOf(ResponseEntity<UserDto> responseEntity) {
    if (responseEntity == null) {
      throw new IllegalStateException("No response from REST service");
    }

    if (isNot2xxSuccessful(responseEntity.getStatusCode())) {
      var badConfiguredResponseMesssage = String.format("Bad configuration of consumer! ResponseCode: %s(%d)",
          responseEntity.getStatusCode().name(),
          responseEntity.getStatusCodeValue()
      );

      throw new IllegalStateException(badConfiguredResponseMesssage);
    }

    return responseEntity.getBody();
  }

  private boolean isNot2xxSuccessful(HttpStatus statusCode) {
    return !statusCode.is2xxSuccessful();
  }
}

File: src/main/kotlin/com/github/jactor/web/consumer/UserConsumer.kt
package com.github.jactor.web.consumer

import com.github.jactor.shared.dto.UserDto
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.util.Optional

@Service
class DefaultUserConsumer(private val restTemplate: RestTemplate) : UserConsumer {

    override fun find(username: String): Optional<UserDto> {
        return Optional.ofNullable(bodyOf(restTemplate.getForEntity("/user/name/$username", UserDto::class.java)))
    }

    override fun findAllUsernames(): List<String> {
        return restTemplate.getForEntity("/user/usernames", Array<String>::class.java).body?.toList() ?: emptyList()
    }

    private fun bodyOf(responseEntity: ResponseEntity<UserDto>?): UserDto? {
        checkNotNull(responseEntity) { "No response from REST service" }

        if (isNot2xxSuccessful(responseEntity.statusCode)) {
            val badConfiguredResponseMesssage = String.format(
                "Bad configuration of consumer! ResponseCode: %s(%d)",
                responseEntity.statusCode.name,
                responseEntity.statusCodeValue
            )
            throw IllegalStateException(badConfiguredResponseMesssage)
        }

        return responseEntity.body
    }

    private fun isNot2xxSuccessful(statusCode: HttpStatus): Boolean {
        return !statusCode.is2xxSuccessful
    }
}

interface UserConsumer {
    fun find(username: String): Optional<UserDto>
    fun findAllUsernames(): List<String>
}
